src/graph/immutable_graph.cc

```diff
@@ -820,85 +818,73 @@ SampledSubgraph ImmutableGraph::SampleSubgraph(IdArray seed_arr,
-      LOG(WARNING)
-        << "The sampling is truncated because we have reached the max number of vertices\n"
-        << "Please use a smaller number of seeds or a small neighborhood";
-      break;
+  for (size_t layer_id = 1; layer_id < num_hops; layer_id++) {
+    sub_ver_map.clear();
+    // sub_vers is used both as a node collection and a queue.
+    // In the while loop, we iterate over sub_vers and new nodes are added to the vector.
+    // A vertex in the vector only needs to be accessed once. If there is a vertex behind idx
+    // isn't in the last level, we will sample its neighbors. If not, the while loop terminates.
+    while (idx < sub_vers.size() && layer_id - 1 == sub_vers[idx].second) {
+      dgl_id_t dst_id = sub_vers[idx].first;
+      int cur_node_level = sub_vers[idx].second;
+      idx++;
+
+      tmp_sampled_src_list.clear();
+      tmp_sampled_edge_list.clear();
+      dgl_id_t ver_len = *(indptr+dst_id+1) - *(indptr+dst_id);
+      if (probability == nullptr) {  // uniform-sample
+        GetUniformSample(val_list + *(indptr + dst_id),
+                         col_list + *(indptr + dst_id),
+                         ver_len,
+                         num_neighbor,
+                         &tmp_sampled_src_list,
+                         &tmp_sampled_edge_list,
+                         &time_seed);
+      } else {  // non-uniform-sample
+        GetNonUniformSample(probability,
+                            val_list + *(indptr + dst_id),
+                            col_list + *(indptr + dst_id),
+                            ver_len,
+                            num_neighbor,
+                            &tmp_sampled_src_list,
+                            &tmp_sampled_edge_list,
+                            &time_seed);
+      }
+      CHECK_EQ(tmp_sampled_src_list.size(), tmp_sampled_edge_list.size());
+      size_t pos = neighbor_list.size();
+      neigh_pos.emplace_back(dst_id, pos);
+      // First we push the size of neighbor vector
+      neighbor_list.push_back(tmp_sampled_edge_list.size());
+      // Then push the vertices
+      for (size_t i = 0; i < tmp_sampled_src_list.size(); ++i) {
+        neighbor_list.push_back(tmp_sampled_src_list[i]);
+      }
+      // Finally we push the edge list
+      for (size_t i = 0; i < tmp_sampled_edge_list.size(); ++i) {
+        neighbor_list.push_back(tmp_sampled_edge_list[i]);
+      }
+      num_edges += tmp_sampled_src_list.size();
+      for (size_t i = 0; i < tmp_sampled_src_list.size(); ++i) {
+        // We need to add the neighbor in the hashtable here. This ensures that
+        // the vertex in the queue is unique. If we see a vertex before, we don't
+        // need to add it to the queue again.
+        auto ret = sub_ver_map.insert(tmp_sampled_src_list[i]);
+        // If the sampled neighbor is inserted to the map successfully.
+        if (ret.second) {
+          sub_vers.emplace_back(tmp_sampled_src_list[i], cur_node_level + 1);
```

Comment:
I don't know how to describe it in words, but here's my understanding of sampling duplicate nodes based on the current implementation:

Consider the following bidirectional graph:
```
C     E
|\   /
| A-B
|/   \
D     F
```
and we are sampling with 3 hops, one neighbor per hop, and the seed node being A.

If in the first hop C is enqueued, then for the second hop, *A can still be sampled, even if A is already in the subgraph*.
If indeed A is sampled, then since A already exists in the nodes subset, *no new neighbors will be sampled from A if A is being sampled again*.  So in the example above, the sampling algorithm will stop there.

Please correct me if I'm wrong.  Otherwise I think this particular property is very subtle, and we need to document it somewhere (perhaps in the Python API).

Reply:
![sample](https://user-images.githubusercontent.com/70481/51471984-2b0af280-1db3-11e9-89fc-1172c254c315.png)
What you described isn't correct. For a graph in (a), we can sample a subgraph where the same node is sampled in different layers (b). When a node is resampled, we can continue sampling from this node.
We have an optimization here, as shown in (c). If a node is sampled multiple times in the same layer, we only sample it once. We could sample the node multiple times and duplicate the node in the layer. However, we feel this is unnecessary. The main goal of having a node in a layer is to estimate its node embedding. Duplicating the node doesn't really help us estimate the node embedding.
The same node can appear in different layer because it has different functions in the GNN model. A node in different layer estimates node embeddings in different layers.

The implementation of the sampling algorithm needs to make sure the sampling probability is strictly followed.

You are right, we should document this behavior clearly.

Reply:
But in line 877-879, if a node already exist in `sub_ver_map`, then `ret.second` will be `false`.  So the item `(A, cur_node_level + 1)` won't be enqueued, and the neighbors of A won't be sampled again?

Reply:
this is to avoid resample the same node in the same layer. `sub_ver_map` is cleared at the beginning when we get to a new layer.